File: tests/fixtures/hook_tests/hooks/live.lua
--- Live event filter functions.
local M = {}

--- Filter: allow create and update, suppress delete.
function M.filter_published(ctx)
    return ctx.operation ~= "delete"
end

--- before_broadcast hook that transforms data.
function M.transform_broadcast(ctx)
    if ctx.data then
        ctx.data._broadcast_marker = "transformed"
    end
    return ctx
end

--- before_broadcast hook that suppresses the event.
function M.suppress_broadcast(ctx)
    return nil
end

--- before_broadcast hook that overrides the title in the event payload.
--- Used to verify that data mutations in before_broadcast affect the event
--- payload but NOT the stored document.
function M.mutate_title_for_broadcast(ctx)
    if ctx.data then
        ctx.data.title = "mutated"
    end
    return ctx
end

--- Filter that gates on the document id and the editing user — verifies that
--- `ctx.document_id` and `ctx.edited_by` reach the live filter.
function M.gate_on_id_and_editor(ctx)
    if ctx.document_id ~= "broadcast-me" then return false end
    if ctx.edited_by == nil then return false end
    return ctx.edited_by.email == "editor@x.com"
end

return M

File: src/hooks/lifecycle/context/live.rs
//! Context passed to a collection's `live.filter` broadcast-gate function.

use serde::Serialize;
use serde_json::Value;

use crate::core::{DocumentFields, event::EventUser};
use crate::typegen::lua::LuaAnnotation;

/// Context passed to the `live.filter` function — the per-collection gate that
/// decides whether a committed mutation is broadcast to live subscribers.
///
/// Runs **after** the write commits, with no CRUD/transaction access. The
/// function returns `true` to broadcast, `false`/`nil` to suppress.
#[derive(Serialize, LuaAnnotation)]
#[lua(class = "crap.LiveFilterContext")]
pub struct LiveFilterContext<'a> {
    /// Collection (or global) slug the mutation targets.
    pub collection: &'a str,
    /// The mutation operation: `"create"`, `"update"`, `"delete"`,
    /// `"undelete"`, `"unpublish"`, or `"restore"`.
    pub operation: &'a str,
    /// The document's field data for this event.
    #[lua(ty = "table<string, any>")]
    pub data: &'a DocumentFields,
    /// The affected document's id, exposed to Lua as `ctx.id` (consistent with
    /// the other hook contexts; the serialized event payload uses `document_id`).
    #[serde(rename = "id")]
    #[lua(rename = "id")]
    pub document_id: &'a str,
    /// The user who caused the mutation (`nil` for anonymous/system changes).
    #[serde(skip_serializing_if = "Option::is_none")]
    #[lua(ty = "{ id: string, email: string }", optional)]
    pub edited_by: Option<&'a EventUser>,
    /// Per-config options from the filter's `{ ref, options }` table; `nil` when
    /// configured as a bare ref string.
    #[serde(skip_serializing_if = "Option::is_none")]
    #[lua(ty = "table", optional)]
    pub options: Option<&'a Value>,
}
